pkg/csi/cinder/controllerserver.go

```diff
@@ -221,5 +221,88 @@ func (cs *controllerServer) ListVolumes(ctx context.Context, req *csi.ListVolume
 	return &csi.ListVolumesResponse{
 		Entries: ventries,
 	}, nil
+}
+
+func (cs *controllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshotRequest) (*csi.CreateSnapshotResponse, error) {
+	// Get OpenStack Provider
+	cloud, err := openstack.GetOpenStackProvider()
+	if err != nil {
+		glog.V(3).Infof("Failed to GetOpenStackProvider: %v", err)
+		return nil, err
+	}
+
+	name := req.Name
+	volumeId := req.SourceVolumeId
+	// No description from csi.CreateSnapshotRequest now
+	description := ""
+
+	// Delegate the check to openstack itself
+	snap, err := cloud.CreateSnapshot(name, volumeId, description, &req.Parameters)
```

Comment:
1. various checks missing here, like if CO request to create a snapshot that already exists, It should not respond any error and return the details of existing snapshot.
2. Snapshot with same name on two different volume-id not allowed.

If you prefer to implement this checks later, add in todo as
TODO: Idempotency Checks

Reply:
@jichenjc for q2 i know it is allowed on openstack side, but after reading csi specs, i think it is not allowed, for q1 you should not respond with any error [It is clearly mentioned in csi specs]

Reply:
ok, I will add more to do and some checks here..